### PR TITLE
codecov reporting

### PR DESCRIPTION
**Pull request by outofambit**
This is a minimal first step at tracking code coverage (#5511). This PR sets up code coverage metrics to be tracked in [codecov](https://codecov.io/gh/desktop/desktop), but doesn't integrate it into our flow (for the vast majority of cases). Codecov will send a status on PRs, but it should always be green unless the coverage drops dramatically. You can click "details" to see the report.

I'd like to let this sit for a bit while we/I think about what ways we'd like to use this tool going forward.

To Do:

setup up codecov reporting on:
- [x] circleci
- [x] appveyor
- [x] travis

finally:
- [x] make sure reports are merged properly

### CODE REVIEW

File: .codecov.yml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+codecov:
+  notify:
+    require_ci_to_pass: yes
+
+coverage:
+  precision: 1
+  round: nearest
+  range: '10...100'
+
+  status:
```
Comment (outofambit) [inline .codecov.yml]: configuration for statuses from codecov:
- only on prs
- only fails if coverage drops by 25%

File: .codecov.yml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+codecov:
+  notify:
+    require_ci_to_pass: yes
+
+coverage:
+  precision: 1
+  round: nearest
+  range: '10...100'
```
Comment (outofambit) [inline .codecov.yml]: this just configures the color threshholds

File: .codecov.yml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  status:
+    project:
+      default:
+        target: auto
+        threshhold: 25
+        base: pr
+    patch: no
+    changes: no
+
+comment: off
```
Comment (outofambit) [inline .codecov.yml]: *no* commenting by codecov with coverage data on PRs

Comment (shiftkey) [reply .codecov.yml]: Do you think this is something we'll want to turn on down the track? What does the commenting look like?

File: .codecov.yml (hunk 2 above)
Comment (shiftkey) [reply .codecov.yml]: Maybe throw a YAML comment in there if there's details that will help the next time we're in here making changes?

File: .codecov.yml (hunk 3 above)
Comment (outofambit) [reply .codecov.yml]: perhaps, once we tune our thresholds and whatnot! it looks like this: https://docs.codecov.io/docs/pull-request-comments

File: .codecov.yml (hunk 2 above)
Comment (outofambit) [reply .codecov.yml]: f238bfd

**Comment by shiftkey** [comment]
> I'd like to let this sit for a bit while we/I think about what ways we'd like to use this tool going forward.

I have two general questions about this PR:

 - what happens after we approve and merge this PR?
 - what outstanding questions around code coverage you have in mind that you'd like to figure out beyond this PR (or get feedback on from others)?

**Comment by outofambit** [comment]
@shiftkey 
> what happens after we approve and merge this PR?

we start to get metrics for code coverage for new branches off of master. we'll get statuses from codecov too on new prs.

> what outstanding questions around code coverage you have in mind that you'd like to figure out beyond this PR (or get feedback on from others)?

- what level of integration do we want with code coverage in our prs?
- it looks like there's a way to focus feedback and track specific parts of the codebase in codecov. is that something that would be more useful than a single number?
- what is a good "rule of thumb" number for enough coverage?

**Comment by shiftkey** [comment]
@outofambit I'm 👍 on both of these things, and suggest using a `meta` issue so we have a place to discuss this further.

**Comment by outofambit** [comment]
@shiftkey 👍 does #5511 seem ok for that?

**Comment by shiftkey** [comment]
@outofambit yep, that works for me!

**Pull request merged**